runtime/io/common/src/aws/smithy/kotlin/runtime/io/internal/JobChannel.kt

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.io.internal
+
+import aws.smithy.kotlin.runtime.io.SdkByteChannel
+import aws.smithy.kotlin.runtime.util.InternalApi
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.Job
+
+/**
+ * Ties the lifetime of the channel and a Job together. Channel failures (closed with exception) will
+ * cause the underlying [Job] to fail with the channel exception.
+ *
+ * This is useful for launching coroutines with the soul purpose of reading/writing data to the channel where
+ * the coroutine should only continue if the channel hasn't failed.
+ */
+@InternalApi
+public class JobChannel(
```

Comment:
**Question**: This only propagates lifecycle events from the channel to the job, not vice versa. Is that important? Would it be possible/correct to also propagate lifecycle from the job to the channel in this class?

Reply:
It depends on the coroutine but I suppose we might be able to. Example [use](https://github.com/awslabs/smithy-kotlin/blob/323dc3d690e34b80538cb88b96eeeefc609f4027/runtime/io/jvm/src/aws/smithy/kotlin/runtime/io/SdkSourceJVM.kt#L36) where we catch all exceptions and forward to the channel. 

An alternative would be to wire up completion handlers:

```kotlin
job.invokeOnCompletation { cause ->
    ch.cancel(cause)
}
```

I don't necessarily see an issue with this (although given the way backround threads/coroutines work in Android I don't know that we want exceptions to bubble up to coroutine exception handlers anyway). Let me think on this more, we can always add it later if we need it. We don't have a lot of use cases for this yet.